WebExtensions対応

fujimotos opened the pull request:
WebExtensions対応版の「Reload on Idle」アドオンを作成しました。

 - 基本的な実装内容は issue #1 の方針に従っています。
 - 設定項目などは可能な限りXPCOM版の実装を踏襲してます。
 - WE版のディレクトリ構成は「Tree Style Tab」アドオンを踏襲しました。

実際にアドオンをFirefox 52.3で試してみて、動作する事を確認してます。

### 対応済みの機能

XPCOM版にあった次のオプションは、このプルリクで実装済みです：

 1. リロード対象タブのURLフィルタ機能
 2. リロード間隔の指定機能 (`idleSeconds`)
 3. ローディング中でもリロードする機能 (`reloadBusyTabs`)

以上はアドオンの設定ページから変更できるようになってます。

### 実装予定の機能

次の機能はこのプルリク後に引き続いて対応します：

 1. 国際化対応 (l10n)
 2. 「編集中のデータがある場合も強制的にリロードする」機能

とくに2の機能については、相談が必要な部分があるので、後ほどissueを立てさせていただきます。

fujimotos committed:
WE: Add WebExtensions-related libraries as submodules.

Add modules for the WebExtension migration:

  1. webextensions-lib-config (Configs.js)
  2. webextensions-lib-options (Options.js)

fujimotos committed:
WE: Create a directory for WebExtension development.

fujimotos committed:
WE: Add the WebExtension port of the background process.

This implements the basic functionality using WebExtensions API.

fujimotos committed:
WE: Define the data structure for addon settings.

This is a straight-forward translation of the XPCOM-based configuration.
(Few options are missing for compatibility reasons, though)

See 'reload-on-idle/content/config.xul' for details.

fujimotos committed:
WE: Implement the setting page.

This commit should add the "Preferences" button to 'about:addon'.

Note that the page layout mimicks the old (XPCOM-based) setting popup.

fujimotos committed:
WE: Create the addon definition (manifest.json)

See: 'reload-on-idle/install.rdf'

fujimotos commented on webextensions/background/background.js:
```diff
@@ -0,0 +1,62 @@
+  });
+};
+
+configs.$loaded.then(main, log);
```
JavaScriptの書き方について補足です：

* 他のアドオンを参考に`Configs.js`と`Options.js`を使って全体を構成してます。
* 既存のコードのスタイルには幾つかバリエーションがあったのですが、
  `TreeStyleTab`に倣って、今回は2スペースで統一してます。
* 主だった関数にはJSDoc形式のドキュメントを付けてます。

あまりPromiseベースの書き方に馴れてないので、変な点があったら何でもご指摘ください。

piroor commented:
反応できてなくてすみません！これから確認します！

piroor commented on webextensions/background/background.js:
```diff
@@ -0,0 +1,62 @@
+/**
+ * Last time we reloaded tabs.
+ */
+var LASTRELOAD = Date.now();
```
全部大文字の識別子はなんとなく定数っぽさを感じます。
値が更新されていく変数の場合、gLastReload のような名前の方が自分はしっくり来ます。
（Mozillaのコーディング規約の「グローバル変数は頭に小文字のgを付ける」に沿った命名）

fujimotos replied:
コミット 40402a9a で変数名を直しました (`LASTRELOAD` → `gLastReload`)

piroor commented on webextensions/background/background.html:
```diff
@@ -0,0 +1,11 @@
+    <script src="/common/Configs.js"></script>
+    <script src="/common/common.js"></script>
+    <script src="/background/background.js"></script>
+  </head>
```
単に複数のスクリプトを読み込むだけで、静的な内容をバックグラウンドページに容易する必要がないのであれば、HTMLファイル無しでmanifest.jsonの「scripts」にこの3つのファイルのパスを記載する方法にしたほうが良さそうな気がします。 （ https://developer.mozilla.org/en-US/Add-ons/WebExtensions/manifest.json/background 参照）

fujimotos replied:
コミット d8aa587 で`scripts`プロパティを使うようにしました。

piroor commented on webextensions/background/background.js:
```diff
@@ -0,0 +1,62 @@
+ * @param {Array} tabs An array of <tabs.Tab>s.
+ * @return {Promise}
+ */
+function reload(tabs) {
```
仮引数名はMozillaのコーディングスタイル https://developer.mozilla.org/ja/docs/Mozilla/Developer_Guide/Mozilla_Coding_Style_Guide に合わせて「argument」の「a」を接頭辞とした「aTabs」のような名前付けにすると、Mozilla関係のソースの文脈によりマッチした感じになるかなと思います。

fujimotos replied:
こちらもコミット 40402a9 でグローバル変数と同じように直しました。
（下の`handleAlarm`関数の変数名も同じように修正しました）

piroor commented on webextensions/background/background.js:
```diff
@@ -0,0 +1,62 @@
+  var reloadBusy = configs.reloadBusyTabs;
+  var chain = [];
+  for (let tab of tabs) {
+    if (!filter.test(tab.url)) continue;
```
ifの条件式のあとは1文だけでも改行したほうがより読みやすくなると思います。（条件式を最後まで読まなくても、次の行に目を向けたときにインデントを伴って「continue」と見えていれば「ああ、ループを継続する条件なんだな」と分かるので）

fujimotos replied:
コミット  a39cced でIF条件文を分割しました。

piroor commented on webextensions/background/background.js:
```diff
@@ -0,0 +1,62 @@
+  var seconds = Math.max(configs.idleSeconds, 15);
+
+  browser.idle.queryState(seconds)
+    .then((state) => {
```
async function handleAlarm...と定義すると、ここでvar state = await browser.idle.queryState(seconds); と書けるようになります。https://developer.mozilla.org/ja/docs/Web/JavaScript/Reference/Statements/async_function コードパスが一本道のケースでは、こちらの書き方の方がネストが浅くなるので読みやすくなるのではないでしょうか？

fujimotos replied:
この部分、コミット 63731e6 で`async/await`を使った書き方に直しました。

piroor commented on webextensions/options/options.html:
```diff
@@ -0,0 +1,38 @@
+  <body>
+    <p><label>
+      <input id="idleSeconds" type="number" min="15">
+      秒以上操作されなかったらタブを再読み込みする
```
このあたりはローカライズの仕組みを使って日英両対応したいところですね！

fujimotos replied:
> このあたりはローカライズの仕組みを使って日英両対応したいところですね！

このMDNの資料に説明のある`i18n`の仕組みですね：

> https://developer.mozilla.org/en-US/Add-ons/WebExtensions/Internationalization

（まだ私がアドオンのローカライズの仕組みを理解できてないので、このプルリクの後で追加PRを提出しようと思います）

piroor commented on webextensions/background/background.js:
```diff
@@ -0,0 +1,62 @@
+'use strict';
```
MPL2.0（1.1もですが）はファイル単位で適用されるライセンスなので、厳密にはライセンスヘッダが各ファイルに必要です。
こちらに各言語向けのサンプルがあります。
https://www.mozilla.org/en-US/MPL/headers/

fujimotos replied:
各JavaScript/HTML/CSSファイルにMPL/2.0のライセンスヘッダを追加しました。

piroor commented:
気付いた点にコメントしてみました。いずれも特に大きな問題というわけではなく、マージ後にこちらで反映するというやり方でも違和感はない程度ですので、参考にしていただければ幸いです。

cosmo0920 commented on webextensions/background/background.js:
```diff
@@ -0,0 +1,62 @@
+
+    var promise = browser.tabs.reload(tab.id);
+    chain.push(promise);
+    log(`reload tab (id=${tab.id}, url=${tab.url})`)
```
文末のセミコロンが欲しいなと思いました。

cosmo0920 replied:
<img width="539" alt="missing-semicolon-after-statement-warnings" src="https://user-images.githubusercontent.com/700876/30677151-f93d4cb2-9ec4-11e7-9141-41a2946aae82.png">

fujimotos replied:
コミット cfbb4b660 でセミコロンを補足しました。

cosmo0920 commented on webextensions/background/background.js:
```diff
@@ -0,0 +1,62 @@
+
+  browser.idle.queryState(seconds)
+    .then((state) => {
+      log(`check state (${state}, ${seconds}s)`)
```
同じく文末のセミコロンが欲しいです

fujimotos replied:
こちらも同じように修正済みです。

cosmo0920 commented:
Emacsのjs2-modeで開いた時に目についた箇所をコメントしてみました。軽微な修正なのでマージには支障がないと思います。

fujimotos commented:
コメントありがとうございました！
いただいた各ポイントについて、今日の往訪から戻り次第、対応いたします。

fujimotos committed:
Add missing semicolons to the end of lines [PR#2]

fujimotos committed:
Make the variable names consistent with the Mozilla coding guide [PR#2]

This fix is an attempt to comfort to the coding standards below:

https://developer.mozilla.org/en-US/docs/Mozilla/Developer_guide/Coding_Style

fujimotos committed:
Add 'MPL/2.0' license header to each file [PR#2]

fujimotos committed:
Break up single-line if conditionals to two lines [PR#2]

fujimotos committed:
Use async/await to avoid deep nesting [PR#2]

fujimotos committed:
Use 'scripts' directive insteads of 'page' property [PR#2]

fujimotos commented:
（上でコメントした「ローカライズ」の部分を除いて）コメントを戴いた各点について対応しました。

piroor commented:
ありがとうございます！
